### PR TITLE
added qlora + fsdp

### PR DESCRIPTION
**Pull request by weifengpy**
```python transformer_nuggets/llama/finetune.py --fsdp_num_gpus 2```

loss goes down on both 1 GPU and 2 GPU: both from 11.2 to 10

next step is reading profiler traces and memory traces

### CODE REVIEW

File: transformer_nuggets/llama/finetune.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+def fsdp_main(rank, world_size, args):
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+    dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    entrypoint(*args, rank=rank, world_size=world_size)
+    dist.destroy_process_group()
+
+
+# copied from https://github.com/pytorch-labs/ao_benchmarks/tree/main/llama
```
Comment (drisspg) [inline transformer_nuggets/llama/finetune.py:152]: lets remove this

Comment (weifengpy) [reply transformer_nuggets/llama/finetune.py:155]: remove the comment itself or avoid copying code?

Comment (weifengpy) [reply transformer_nuggets/llama/finetune.py]: I removed the comment

File: transformer_nuggets/llama/finetune.py [hunk 2]
```diff
@@ -87,18 +123,233 @@ def main(
...
+
+            if training_config.profile and iter_num < 103:
+                # We want to profile iters 100-102 of the model training
+                p.step()
+
+            if training_config.track_max_memory and rank == 0:
+                print(
+                    "iter_num",
+                    iter_num,
+                    "mem usage GB",
```
Comment (drisspg) [inline transformer_nuggets/llama/finetune.py:266]: nit this shoudl actually be GiB

File: transformer_nuggets/llama/finetune.py [hunk 3]
```diff
@@ -58,20 +79,35 @@ def main(
...
+            model,
+            use_orig_params=True,
+            auto_wrap_policy=ModuleWrapPolicy([TransformerBlock]),
+        )
+
+    if training_config.compile:
+        model = torch.compile(model)
+
+    if rank == 0:
+        print(model)
```
Comment (drisspg) [inline transformer_nuggets/llama/finetune.py:109]: nit: log

**Pull request merged**